src/dbnode/persist/fs/registry.go

```diff
@@ -0,0 +1,335 @@
+)
+
+type fileRegistry struct {
+	sync.RWMutex
+
+	entries map[string]*fileStatus
+
+	clockOpts clock.Options
+	logger    *zap.Logger
+}
+
+// NewFileRegistry creates a new file registry.
+func NewFileRegistry(
+	clockOpts clock.Options,
+	instrumentOpts instrument.Options,
+) FileRegistry {
+	return &fileRegistry{
+		entries:   make(map[string]*fileStatus),
+		clockOpts: clockOpts,
+		logger:    instrumentOpts.Logger(),
+	}
+}
+
+func (r *fileRegistry) Open(filePath string) (File, error) {
+	entry, err := r.newFileStatusOrExisting(filePath)
+	if err != nil {
+		return nil, err
+	}
+	return entry.Open()
+}
+
+func (r *fileRegistry) OpenFile(filePath string, flag int, perm os.FileMode) (File, error) {
+	entry, err := r.newFileStatusOrExisting(filePath)
+	if err != nil {
+		return nil, err
+	}
+	return entry.OpenFile(flag, perm)
+}
+
+func (r *fileRegistry) Remove(filePath string) error {
+	pathAbs, err := filepath.Abs(filePath)
+	if err != nil {
+		return err
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	status, ok := r.entries[pathAbs]
+	if !ok {
+		// Just fine, no files open.
+		return os.Remove(pathAbs)
+	}
+
+	// Delete from the registry.
+	delete(r.entries, pathAbs)
+
+	// Notify open file holders that this file is being removed
+	// and perform a delayed remove which will delete the file
+	// after all references are released.
+	go func() {
+		err := status.NotifyAndRemove()
+		if err != nil {
+			r.logger.Error("unable to delayed remove file",
+				zap.String("path", pathAbs), zap.Error(err))
+		}
+	}()
+
+	return nil
+}
+
+func (r *fileRegistry) newFileStatusOrExisting(filePath string) (*fileStatus, error) {
+	pathAbs, err := filepath.Abs(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	r.RLock()
+	status, ok := r.entries[pathAbs]
+	r.RUnlock()
+
+	if ok {
+		return status, nil
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	status, ok = r.entries[pathAbs]
+	if !ok {
+		status = newFileStatus(pathAbs, r.clockOpts.NowFn(), r.logger)
+		r.entries[pathAbs] = status
+	}
+
+	return status, nil
+}
+
+type fileStatus struct {
+	sync.RWMutex
+	status    FileStatus
+	pathAbs   string
+	watchable watch.Watchable
+	refs      []*fileRef
+
+	nowFn clock.NowFn
+
+	logger *zap.Logger
+}
+
+func newFileStatus(pathAbs string, nowFn clock.NowFn, logger *zap.Logger) *fileStatus {
+	f := &fileStatus{
+		status: FileStatus{
+			Type: AvailableFileStatusType,
+		},
+		pathAbs: pathAbs,
+		nowFn:   nowFn,
+		logger:  logger,
+	}
+	watchable := watch.NewWatchable()
+	watchable.Update(f.status)
+	f.watchable = watchable
+	return f
+}
+
+func (f *fileStatus) newFileRef(ref *os.File) *fileRef {
+	fileRef := newFileRef(f, ref)
+	f.Lock()
+	f.refs = append(f.refs, fileRef)
+	f.Unlock()
+	return fileRef
+}
+
+func (f *fileStatus) removeFileRef(elem *fileRef) error {
+	f.Lock()
+	defer f.Unlock()
+
+	if len(f.refs) == 0 {
+		return fmt.Errorf("remove file ref not found for: %v", elem.PathAbs())
+	}
+
+	refs := make([]*fileRef, 0, len(f.refs)-1)
```

Comment:
Think its worth just reslicing and performing a copy? Maybe an over-optimization, but I can imagine this happening in a tight loop when bblocks expire and such